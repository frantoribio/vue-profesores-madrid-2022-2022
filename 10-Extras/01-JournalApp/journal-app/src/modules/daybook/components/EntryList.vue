<template>
  <div class="h-[calc(88vh)] border-r border-primary">
    <div class="mr-2 px-2 pt-2">
      <input
        data-testid="entrylist-input"
        v-model.trim="searchTerm"
        type="text"
        placeholder="Buscar entradas"
        class="input input-bordered input-primary w-full rounded-md bg-base-100 text-base-content"
      />
    </div>
    <div class="mx-2 mt-2 flex justify-center">
      <button
        data-testid="entrylist-new"
        class="btn btn-outline btn-primary"
        @click="goToNewEntry()"
      >
        <Icon
          icon="ci:file-new"
          class="mr-2 h-6 w-6"
        />
        Nueva Entrada
      </button>
    </div>
    <div class="scrollbar mt-1 h-[calc(88vh_-_90px)] overflow-y-auto">
      <div>
        <!-- Animaciones -->
        <transition-group
          name="list-complete"
          tag="EntryItem"
        >
          <EntryItem
            data-testid="entrylist-item"
            v-for="entry in entriesByTerm"
            :key="entry.id"
            :entry="entry"
            @click="goToEntry(entry)"
          />
        </transition-group>
      </div>
    </div>
  </div>
</template>

<script setup>
  import { Icon } from '@iconify/vue'
  import { computed, ref } from 'vue'
  import { useRouter } from 'vue-router'
  import EntryItem from '../components/EntryItem.vue'
  import JournalStore from '../stores/journal'

  const router = useRouter()
  const journalStore = JournalStore()

  const searchTerm = ref('')

  const entriesByTerm = computed(() => journalStore.getEntriesByTerm(searchTerm.value))

  const goToEntry = (entry) => {
    router.push({ name: 'daybook-entry', params: { id: entry.id } })
  }

  const goToNewEntry = () => {
    router.push({ name: 'daybook-entry', params: { id: 'new' } })
  }
</script>

<style lang="scss" scoped>
  // https://vuejs.org/guide/built-ins/transition-group.html
  .list-complete-item {
    transition: all 0.8s ease;
  }
  .list-complete-enter-from,
  .list-complete-leave-to {
    opacity: 0;
    transform: translateY(30px);
  }
  .list-complete-leave-active {
    position: relative;
  }
</style>
