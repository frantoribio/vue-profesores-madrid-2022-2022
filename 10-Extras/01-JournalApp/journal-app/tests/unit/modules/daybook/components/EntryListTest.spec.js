import { beforeEach, describe, expect, test } from 'vitest'

import EntryList from '@/modules/daybook/components/EntryList.vue'

import { shallowMount } from '@vue/test-utils'

// Suite de Test
// Necesitamos Pinia para que funcione el test
// Necesitamos router para que funcione el test
import { createTestingPinia } from '@pinia/testing'
import { createRouterMock, injectRouterMock } from 'vue-router-mock'

import entries from '../stores/mocks/entries.mocks'

describe('Daybook Componente -> EntryList', () => {
  let wrapper
  const router = createRouterMock()

  beforeEach(() => {
    injectRouterMock(router)
    wrapper = shallowMount(EntryList, {
      // Le asignamos un estado inicial a la estore mock de prueba JournalStore
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              JournalStore: { entries: entries },
            },
          }),
        ],
      },
    })
  })

  test('Debe hacer match con el snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('Debe mostrar una lista de entradas', () => {
    const entriesByTerm = wrapper.vm.entriesByTerm
    expect(entriesByTerm).toEqual(entries)
    const entriesListItems = wrapper.findAll('[data-testid="entrylist-item"]')
    expect(entriesListItems.length).toBe(entries.length)
  })

  test('Debe mostrar una lista de entradas filtradas por una palabra clave', async () => {
    const input = wrapper.find('[data-testid="entrylist-input"]')
    input.setValue('Lorem ipsum dolor sit amet, consectetur adipiscing elit')
    await input.trigger('input')
    const entriesByTerm = wrapper.vm.entriesByTerm
    expect(entriesByTerm).toEqual([entries[0]])
    const entriesListItems = wrapper.findAll('[data-testid="entrylist-item"]')
    expect(entriesListItems.length).toBe(1)
  })

  test('Al pulsar el botÃ³n de nuevo debe ir a nueva entrada', async () => {
    const button = wrapper.find('[data-testid="entrylist-new"]')
    await button.trigger('click')
    expect(router.push).toHaveBeenCalledWith({ name: 'daybook-entry', params: { id: 'new' } })
  })

  test('Al un item de la lista debe ir a la vista de esa entrada', async () => {
    // Cogemos el primero de la lista
    const entryListItem = wrapper.find('[data-testid="entrylist-item"]')
    await entryListItem.trigger('click')
    expect(router.push).toHaveBeenCalledWith({
      name: 'daybook-entry',
      params: { id: entries[0].id },
    })
  })
})
