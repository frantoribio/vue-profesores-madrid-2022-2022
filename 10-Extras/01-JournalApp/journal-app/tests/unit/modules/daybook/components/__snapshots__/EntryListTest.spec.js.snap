// Vitest Snapshot v1

exports[`Daybook Componente -> EntryList > Debe hacer match con el snapshot 1`] = `
"<div class=\\"h-[calc(88vh)] border-r border-primary\\" data-v-d2f66aca=\\"\\">
  <div class=\\"mr-2 px-2 pt-2\\" data-v-d2f66aca=\\"\\"><input data-testid=\\"entrylist-input\\" type=\\"text\\" placeholder=\\"Buscar entradas\\" class=\\"input input-bordered input-primary w-full rounded-md bg-base-100 text-base-content\\" data-v-d2f66aca=\\"\\"></div>
  <div class=\\"mx-2 mt-2 flex justify-center\\" data-v-d2f66aca=\\"\\"><button data-testid=\\"entrylist-new\\" class=\\"btn btn-outline btn-primary\\" data-v-d2f66aca=\\"\\">
      <icon-stub icon=\\"ci:file-new\\" class=\\"mr-2 h-6 w-6\\" data-v-d2f66aca=\\"\\"></icon-stub> Nueva Entrada
    </button></div>
  <div class=\\"scrollbar mt-1 h-[calc(88vh_-_90px)] overflow-y-auto\\" data-v-d2f66aca=\\"\\">
    <div data-v-d2f66aca=\\"\\">
      <!-- Animaciones -->
      <transition-group-stub data-v-d2f66aca=\\"\\">
        <entry-item-stub entry=\\"[object Object]\\" data-testid=\\"entrylist-item\\" data-v-d2f66aca=\\"\\"></entry-item-stub>
        <entry-item-stub entry=\\"[object Object]\\" data-testid=\\"entrylist-item\\" data-v-d2f66aca=\\"\\"></entry-item-stub>
        <entry-item-stub entry=\\"[object Object]\\" data-testid=\\"entrylist-item\\" data-v-d2f66aca=\\"\\"></entry-item-stub>
      </transition-group-stub>
    </div>
  </div>
</div>"
`;
